src/sherpa_ai/agents/base.py

```diff
@@ -119,35 +125,85 @@ def validate_output(self):
         Returns:
             str: The synthesized output after validation.
         """
-        failed_validation = []
-        result = self.synthesize_output()
+        result = ""
+        # create array of instance of validation so that we can keep track of how many times regeneration happened.
+        instantiated_validations = []
         for validation in self.validations:
-            for count in range(self.validation_steps):
-                self.belief.update_internal(EventType.result, self.name, result)
+            try:
+                instantiated_validations.append(validation())
+            except Exception as e:
+                instantiated_validations.append(validation)
+
+        all_pass = False
+        validation_is_scaped = False
+        while_count = 0
+
+        # this loop will run until max regeneration reached or all validations have failed
+        while (
+            self.global_regen_max > sum(val.count for val in instantiated_validations)
+            and not all_pass
+        ):
+            logger.info(f"validations_size: {len(instantiated_validations)}")
+            break_while = False
+            if break_while:
+                break
+            while_count += 1
+            logger.info(f"main_iteration: {while_count}")
+            logger.info(
+                f"regen_count: {sum(val.count for val in instantiated_validations)}"
+            )
+            for x in range(len(instantiated_validations)):
+                validation = instantiated_validations[x]
+                logger.info(f"validation_running: {validation.__class__.__name__}")
+                logger.info(f"validation_count: {validation.count}")
+                if validation.count < self.validation_steps:
```

Comment:
Shouldn't the `validation_steps` associated with each validation?

Reply:
it does the validation.count is the number of times it run

Reply:
Yeah, this part makes sense. I'm thinking that in the design, we ideally want to be able to set a bound on the validation steps differently for each validator, isn't this the case here?